File: apps/tmjpad/tmjpad/persistence.py
"""Session and buffer persistence for TMJPad.

Each open tab has a buffer file at ~/.config/tmjpad/buffers/<uuid>.txt that
mirrors the current text. The session.json file holds tab list (order, paths,
cursor positions, active tab, window size).

Auto-save writes to buffer files on every modification (debounced). The
session.json is rewritten on tab open/close/save/reorder.

Goal: zero data loss even on hard crash. Source of truth on restart is the
buffer files; session.json describes how to display them.

This module is pure Python (no GTK) so it's easily unit-testable.
"""
from __future__ import annotations

import json
import os
import uuid
from dataclasses import asdict, dataclass, field
from pathlib import Path


def _config_dir() -> Path:
    base = os.environ.get("XDG_CONFIG_HOME") or str(Path.home() / ".config")
    return Path(base) / "tmjpad"


CONFIG_DIR = _config_dir()
BUFFERS_DIR = CONFIG_DIR / "buffers"
SESSION_FILE = CONFIG_DIR / "session.json"


@dataclass
class TabState:
    """Persistent state of a single tab."""

    id: str
    title: str
    path: str | None  # filesystem path; None for untitled
    cursor_offset: int = 0

    def buffer_path(self, base: Path = BUFFERS_DIR) -> Path:
        return base / f"{self.id}.txt"


@dataclass
class Session:
    tabs: list[TabState] = field(default_factory=list)
    active_index: int = 0
    window_width: int = 1100
    window_height: int = 700

    @classmethod
    def load(cls, session_file: Path = SESSION_FILE) -> Session:
        """Read session.json. Returns empty Session if missing or corrupt.

        On corruption, the bad file is renamed to .json.bak so debug is possible.
        """
        if not session_file.exists():
            return cls()
        try:
            raw = json.loads(session_file.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            session_file.replace(session_file.with_suffix(".json.bak"))
            return cls()
        try:
            tabs = [TabState(**t) for t in raw.get("tabs", [])]
            return cls(
                tabs=tabs,
                active_index=raw.get("active_index", 0),
                window_width=raw.get("window_width", 1100),
                window_height=raw.get("window_height", 700),
            )
        except (TypeError, KeyError):
            session_file.replace(session_file.with_suffix(".json.bak"))
            return cls()

    def save(self, session_file: Path = SESSION_FILE) -> None:
        """Atomic write: tmp file + rename."""
        session_file.parent.mkdir(parents=True, exist_ok=True)
        BUFFERS_DIR.mkdir(parents=True, exist_ok=True)
        tmp = session_file.with_suffix(".json.tmp")
        payload = {
            "tabs": [asdict(t) for t in self.tabs],
            "active_index": self.active_index,
            "window_width": self.window_width,
            "window_height": self.window_height,
        }
        tmp.write_text(json.dumps(payload, indent=2), encoding="utf-8")
        tmp.replace(session_file)


def new_tab_state(title: str = "Untitled-1", path: str | None = None) -> TabState:
    return TabState(id=str(uuid.uuid4()), title=title, path=path)


def write_buffer(state: TabState, content: str, base: Path = BUFFERS_DIR) -> None:
    """Atomic write of buffer content."""
    base.mkdir(parents=True, exist_ok=True)
    buf = state.buffer_path(base)
    tmp = buf.with_suffix(".tmp")
    tmp.write_text(content, encoding="utf-8")
    tmp.replace(buf)


def read_buffer(state: TabState, base: Path = BUFFERS_DIR) -> str:
    buf = state.buffer_path(base)
    if buf.exists():
        return buf.read_text(encoding="utf-8")
    return ""


def remove_buffer(state: TabState, base: Path = BUFFERS_DIR) -> None:
    buf = state.buffer_path(base)
    if buf.exists():
        buf.unlink()


def next_untitled_title(existing_titles: set[str]) -> str:
    """Generate next available 'Untitled-N' name not in existing_titles."""
    n = 1
    while f"Untitled-{n}" in existing_titles:
        n += 1
    return f"Untitled-{n}"

File: apps/tmjpad/src/persistence.rs
//! Session and buffer persistence.
//!
//! Cada tab tem um buffer file em `~/.config/tmjpad/buffers/<uuid>.txt`
//! que espelha o texto atual. session.json contém lista de tabs (ordem,
//! paths, cursor positions, active tab, window size).
//!
//! Auto-save escreve nos buffer files em cada modificação (debounced).
//! session.json é re-escrito em open/close/save/reorder de tab.
//!
//! Source of truth no restart: os buffer files. session.json descreve
//! como mostrar.

use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use std::{fs, io};
use uuid::Uuid;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct TabState {
    pub id: String,
    pub title: String,
    pub path: Option<String>, // None = untitled
    #[serde(default)]
    pub cursor_offset: i32,
}

impl TabState {
    pub fn new(title: impl Into<String>, path: Option<String>) -> Self {
        Self {
            id: Uuid::new_v4().to_string(),
            title: title.into(),
            path,
            cursor_offset: 0,
        }
    }

    pub fn buffer_path(&self, base: &Path) -> PathBuf {
        base.join(format!("{}.txt", self.id))
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Session {
    #[serde(default)]
    pub tabs: Vec<TabState>,
    #[serde(default)]
    pub active_index: usize,
    #[serde(default = "default_width")]
    pub window_width: i32,
    #[serde(default = "default_height")]
    pub window_height: i32,
}

fn default_width() -> i32 {
    1100
}

fn default_height() -> i32 {
    700
}

impl Default for Session {
    fn default() -> Self {
        Self {
            tabs: Vec::new(),
            active_index: 0,
            window_width: default_width(),
            window_height: default_height(),
        }
    }
}

pub fn config_dir() -> PathBuf {
    dirs::config_dir()
        .unwrap_or_else(|| PathBuf::from(".config"))
        .join("tmjpad")
}

pub fn buffers_dir() -> PathBuf {
    config_dir().join("buffers")
}

pub fn session_file() -> PathBuf {
    config_dir().join("session.json")
}

impl Session {
    /// Lê session.json. Retorna `Session::default()` se ausente ou corrupto.
    /// Em caso de corrupção, renomeia o arquivo ruim pra `.json.bak`.
    pub fn load() -> Self {
        let session_path = session_file();
        if !session_path.exists() {
            return Self::default();
        }

        let raw = match fs::read_to_string(&session_path) {
            Ok(s) => s,
            Err(_) => return Self::default(),
        };

        match serde_json::from_str::<Session>(&raw) {
            Ok(s) => s,
            Err(_) => {
                let bak = session_path.with_extension("json.bak");
                let _ = fs::rename(&session_path, &bak);
                Self::default()
            }
        }
    }

    /// Atomic write: arquivo tmp + rename.
    pub fn save(&self) -> io::Result<()> {
        let session_path = session_file();
        if let Some(parent) = session_path.parent() {
            fs::create_dir_all(parent)?;
        }
        fs::create_dir_all(buffers_dir())?;

        let tmp = session_path.with_extension("json.tmp");
        let json = serde_json::to_string_pretty(self)
            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
        fs::write(&tmp, json)?;
        fs::rename(&tmp, &session_path)?;
        Ok(())
    }
}

/// Gera próximo nome 'Untitled-N' não usado.
pub fn next_untitled_title<I: IntoIterator<Item = S>, S: AsRef<str>>(existing: I) -> String {
    let existing: std::collections::HashSet<String> =
        existing.into_iter().map(|s| s.as_ref().to_string()).collect();
    let mut n = 1;
    loop {
        let candidate = format!("Untitled-{n}");
        if !existing.contains(&candidate) {
            return candidate;
        }
        n += 1;
    }
}

/// Atomic write do buffer file.
pub fn write_buffer(state: &TabState, content: &str) -> io::Result<()> {
    let base = buffers_dir();
    fs::create_dir_all(&base)?;
    let buf = state.buffer_path(&base);
    let tmp = buf.with_extension("tmp");
    fs::write(&tmp, content)?;
    fs::rename(&tmp, &buf)?;
    Ok(())
}

pub fn read_buffer(state: &TabState) -> String {
    let buf = state.buffer_path(&buffers_dir());
    fs::read_to_string(&buf).unwrap_or_default()
}

pub fn remove_buffer(state: &TabState) {
    let buf = state.buffer_path(&buffers_dir());
    let _ = fs::remove_file(&buf);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn next_untitled_with_empty() {
        let empty: Vec<&str> = vec![];
        assert_eq!(next_untitled_title(empty), "Untitled-1");
    }

    #[test]
    fn next_untitled_skips_existing() {
        assert_eq!(
            next_untitled_title(vec!["Untitled-1", "Untitled-2"]),
            "Untitled-3"
        );
    }

    #[test]
    fn session_default_dimensions() {
        let s = Session::default();
        assert_eq!(s.window_width, 1100);
        assert_eq!(s.window_height, 700);
    }

    #[test]
    fn tab_state_new_has_uuid() {
        let t = TabState::new("foo", None);
        assert!(!t.id.is_empty());
        assert_eq!(t.title, "foo");
        assert!(t.path.is_none());
    }
}
